### ISSUE TITLE
mlr3learners not working with mlr3pipelines

### ISSUE THREAD

**Issue by jonnybaik** [mlr-org/mlr3pipelines#287]
I am following the introduction to mlr3pipelines [here](https://mlr3pipelines.mlr-org.com/articles/introduction.html#pipeoplearner). The example classification pipeline works, but it fails when I try to create a similar pipeline using a learner from mlr3learners.

The example works:

```r
library(mlr3)
library(mlr3pipelines)

# A simple classification example, works as expected
task = TaskClassif$new("iris", as_data_backend(iris), "Species")

gr = mlr_pipeops$get("pca") %>>% 
  mlr_pipeops$get("learner",
                  mlr_learners$get("classif.rpart"))

gr$train(task)
# $classif.rpart.output
# NULL

gr$predict(task)
# $classif.rpart.output
# <PredictionClassif> for 150 observations:
#     row_id     truth  response
#          1    setosa    setosa
#          2    setosa    setosa
#          3    setosa    setosa
# ---                           
#        148 virginica virginica
#        149 virginica virginica
#        150 virginica virginica
```

But try using a learner from the `mlr3learners` package:

```r
library(mlr3learners)
library(xgboost)

task = TaskClassif$new("iris", as_data_backend(iris), "Species")

gr = mlr_pipeops$get("pca") %>>%
  mlr_pipeops$get("learner",
                  mlr_learners$get("classif.xgboost"))
gr$train(task)
# Error in learner_train(self, task, row_ids) : 
#   could not find function "learner_train"
```

I tried to create a simple regression pipeline as well using mlr3's provided `regr.lm` learner:

```r
task = TaskRegr$new("mtcars", as_data_backend(mtcars), "mpg")

gr = mlr_pipeops$get("pca") %>>%
  mlr_pipeops$get("learner",
                  mlr_learners$get("regr.rpart"))
gr$train(task)
# Error in learner_train(self, task, row_ids) : 
#   could not find function "learner_train"
```

The `regr.lm` works outside of the pipeline.

```r
task = TaskRegr$new("mtcars", as_data_backend(mtcars), "mpg")

learner = mlr_learners$get("regr.lm")
learner$train(task)
learner$predict(task)
# <PredictionRegr> for 32 observations:
#                 row_id truth response
#            AMC Javelin  15.2 17.73218
#     Cadillac Fleetwood  10.4 12.03401
#             Camaro Z28  13.3 13.30602
# ---                                  
#          Toyota Corona  21.5 23.64310
#                Valiant  18.1 20.38304
#             Volvo 142E  21.4 24.36827
```

For now, I am working around this by not including the learner in the preprocessing pipeline and keeping the preprocessing and training separate. I can still perform preprocessing and then pass along the task to the learner just fine.

```r
task = TaskRegr$new("mtcars", as_data_backend(mtcars), "mpg")

pp = mlr_pipeops$get("scalerange") %>>%
  mlr_pipeops$get("pca", param_vals = list(center = FALSE, scale. = FALSE))

learner = mlr_learners$get("regr.lm")

# Train the preprocessor
pp$train(task)
# Preprocess the data and grab the task
pp_task <- pp$predict(task)[[1]]
# Train the model
learner$train(pp_task)
# Get predictions
learner$predict(pp_task)
# <PredictionRegr> for 32 observations:
#                 row_id truth response
#            AMC Javelin  15.2 17.73218
#     Cadillac Fleetwood  10.4 12.03401
#             Camaro Z28  13.3 13.30602
# ---                                  
#          Toyota Corona  21.5 23.64310
#                Valiant  18.1 20.38304
#             Volvo 142E  21.4 24.36827
```

**Comment by mllg** [comment]
I cannot reproduce this. 

Maybe the package versions are not compatible? I tried with the CRAN versions of mlr3, mlr3pipelines and mlr3learners.

@mb706 Can you try?

**Comment by jonnybaik** [comment]
My session info, if it helps:

```r
R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS: /usr/lib/atlas-base/libf77blas.so.3.0
LAPACK: /usr/lib/lapack/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_CA.UTF-8       LC_NUMERIC=C               LC_TIME=en_CA.UTF-8        LC_COLLATE=en_CA.UTF-8     LC_MONETARY=en_CA.UTF-8   
 [6] LC_MESSAGES=en_CA.UTF-8    LC_PAPER=en_CA.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_CA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xgboost_0.90.0.2    mlr3learners_0.1.3  mlr3pipelines_0.1.0 mlr3_0.1.3         

loaded via a namespace (and not attached):
 [1] lgr_0.3.3         paradox_0.1.0     lattice_0.20-35   mlr3misc_0.1.5    packrat_0.5.0     digest_0.6.22     crayon_1.3.4      withr_2.1.2      
 [9] grid_3.5.1        R6_2.4.0          backports_1.1.5   magrittr_1.5      stringi_1.4.3     uuid_0.1-2        data.table_1.12.6 rstudioapi_0.10  
[17] Matrix_1.2-14     rpart_4.1-13      checkmate_1.9.4   tools_3.5.1       compiler_3.5.1   
```

**Comment by berndbischl** [comment]
@jonnybaik what happens if you update all mlr3* packages from cran?
and if that still fails from github?

**Comment by jonnybaik** [comment]
I tried re-installing the packages from CRAN and it did not fix the problem, **but** re-installing the packages with `dependencies = TRUE` did resolve my issue. I am not sure why that fixed my problem. As far as I can see, with `dependencies = TRUE`, R installed/updated the following packages: `‘DiceKriging’, ‘e1071’, ‘kknn’, ‘ranger’`. I don't understand how any of those packages would have interfered with creating a PipeOp learner.

```r
install.packages(c("mlr3", "mlr3pipelines", "mlr3learners"), repos = "cran.rstudio.com", dependencies = TRUE)
# Installing packages into ‘/home/jonnybaik/R/x86_64-pc-linux-gnu-library/3.5’
# (as ‘lib’ is unspecified)
# also installing the dependencies ‘DiceKriging’, ‘e1071’, ‘kknn’, ‘ranger’
# [...]
```

**Comment by mllg** [comment]
Should have solved itself, otherwise we would get swamped with bug reports.

**Comment by mb706** [comment]
Closing this issue, the following is just notes for myself in case I ever want to remember this.

---

So I tried to reproduce this issue and https://github.com/mlr-org/mlr3pipelines/issues/286, which both look similar: a function that is internal to an mlr3-package is reported as not found.

In #286, it is the `%&&%`-function, which was at that point in utils.R (by now renamed `%check&&%`), which was not found by the custom_check function of the `"mutate"`-`PipeOp` (which is a function inside the mlr3pipelines package).
```
 1: mlr_pipeops$get("mutate")
 2: obj$value$new(...)
 3: .subset2(public_bind_env, "initialize")(...)
 4: (function (xs) 
{
    if (missing(xs)) {
        return(private$.values)
  
 5: self$assert(xs)
 6: makeAssertion(xs, self$check(xs), .var.name, NULL)
 7: isTRUE(res)
 8: self$check(xs)
 9: self$params[[n]]$check(xs[[n]])
10: private$.check(x)
11: self$custom_check(x)
12: check_list(x, types = "formula", names = "unique") %&&% Reduce(`%&&%`, lapp
```

In this issue, the problem is that `mlr3:::learner_train` is not found from within `Learner$train()`, when called from `PipeOpLearner`.
```
1: gr$train(task)
2: graph_reduce(self, input, "train", single_input)
3: op[[fun]](input)
4: self$train_internal(input)
5: private$.learner$train(task)
6: learner_train(self, task, row_ids)
```

I compliled the exact R-version (3.5.1) and made sure I had the exact packages loaded as jonnybaik reports, installing them from the mran-snapshot from October 22, 2019. I still can't reproduce these issues. Note how jonnybaik's `sessionInfo()` also didn't change between when things worked and when they didn't.

It could be that something messed up R's internals at some point. What is weird here is that this doesn't appear to be a problem with one specific R-session, since e.g. package version of `digest` changed from 0.6.21 to 0.6.22 between this issue and #286. Maybe it could be something caused by packrat, and a specific setup involving packrat?

It appears this will remain a mystery, unfortunately.
